packages/material-ui/src/transitions/utils.ts

```diff
@@ -0,0 +1,42 @@
+import * as React from 'react';
+
+export const reflow = (node: Element) => node.scrollTop;
+
+interface ComponentProps {
+  easing: string | { enter?: string; exit?: string };
+  style: React.CSSProperties;
+  timeout: number | { enter?: number; exit?: number };
+}
+
+interface Options {
+  mode: 'enter' | 'exit';
+}
+
+interface TransitionProps {
+  duration: number | undefined;
+  easing: string | undefined;
+  delay: string | undefined;
+}
+
+export function getTransitionProps(props: ComponentProps, options: Options): TransitionProps {
+  const { timeout, easing, style = {} } = props;
+
+  return {
+    // TODO: `transitionDuration` implies `timeout: number` but we warn at no point if a user breaks these assumptions
+    // @ts-expect-error
+    duration:
+      style.transitionDuration || typeof timeout === 'number'
```

Comment:
```suggestion
      typeof timeout === 'number'
```
Broke no tests but fixed the invalid types e.g. when using `<Zoom style={{ transitionDuration: '1s' }}` (which [triggers a warning](https://codesandbox.io/s/simplezoom-material-demo-forked-ezy8t?file=/demo.tsx) that doesn't tell that the props are the problem).

It seems to me we just want to return `undefined` if `transitionDuration` is defined? Same issue for `easing`.

@oliviertassinari Any idea why this behavior is important? The function was introduced in https://github.com/mui-org/material-ui/pull/10151 which already introduced a change in behavior in https://github.com/eps1lon/material-ui/commit/98e065e20bfedc52022ed464067b4ae701126b79#diff-04f55dbede7c29675e5ef7ae85442d2c4a6fe92717583aec061b450aa866a449R28-R33

Reply:
The test case you are looking for is this demo: https://next.material-ui.com/components/floating-action-button/#animation

Reply:
I mean, being able to customize the duration from the style prop. We don't rely on this logic anywhere in the codebase. Would the following work?

```diff
diff --git a/packages/material-ui/src/transitions/utils.ts b/packages/material-ui/src/transitions/utils.ts
index 6b56a9d28b..5a384a98b2 100644
--- a/packages/material-ui/src/transitions/utils.ts
+++ b/packages/material-ui/src/transitions/utils.ts
@@ -13,7 +13,7 @@ interface Options {
 }

 interface TransitionProps {
-  duration: number | undefined;
+  duration: string | number | undefined;
   easing: string | undefined;
   delay: string | undefined;
 }
@@ -22,12 +22,10 @@ export function getTransitionProps(props: ComponentProps, options: Options): Tra
   const { timeout, easing, style = {} } = props;

   return {
-    // TODO: `transitionDuration` implies `timeout: number` but we warn at no point if a user breaks these assumptions
-    // @ts-expect-error
     duration:
-      style.transitionDuration || typeof timeout === 'number'
-        ? timeout
-        : timeout[options.mode] || 0,
+      style.transitionDuration ||
+      (typeof timeout === 'number' ? timeout : timeout[options.mode]) ||
+      0,
     easing:
       style.transitionTimingFunction || typeof easing === 'object'
         ? // TODO: Looks odd that we want `undefined` for `transitionTimingFunction` && `easing: string`.
```

Reply:
Yep this makes sense. The way this behavior was encoded was just unreadable.

Reply:
Especially because the new implementation would make sure `style.transitionDuration` is actually used. The old behavior still relied on the call site to make sure that if `options.duration` was returned as `undefined` that `style.transitionDuration` was used ultimately

Reply:
Wait. If `style.transitionDuration` is defined and `timeout` is e.g. `5`. Then we would've returned `5`. But now we return `style.transitionDuration`.

Also: Why default to `0` in duration but `undefined` in `easing`? These values get treated differently in  https://github.com/mui-org/material-ui/blob/5a983eadb806ba095de2a2754b208d470e3f55e7/packages/material-ui/src/styles/createTransitions.js#L59-L61

Reply:
> If style.transitionDuration is defined and timeout is e.g. 5. Then we would've returned 5

From my perspective, the inline should have more priority, 5 should not be returned. Why use inline style if not to have more priority?

> Also: Why default to 0 in duration but undefined in easing? These values get treated differently in

I don't see a use case for the default at 0. 👍 for trying to remove it.

Reply:
> Why use inline style if not to have more priority?

This is the question directed at you. You wrote the previous code.

Reply:
Right, the previous logic code was wrong. My intent was to have `style.transitionDuration` to have the most priority.

Reply:
Gotcha, just want to make sure we're clear about the direction since this is no longer just an internal refactoring but behavioral change.